### ISSUE TITLE
/imsx not visible inside (??{})

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3555]
Migrated from [rt.perl.org#6011](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6011) (status was 'resolved')

Searchable as RT6011$

**Comment by p5pRT** [comment]
### From rick@consumercontact.com




---
$ perl \-le 'print "not ok" if "a\\n"\!~/\.\(??\{"\."\}\)/s'
not ok
$ perl \-le 'print "not ok" if "aA"\!~/a\(??\{"a"\}\)/i'
not ok
 
etc\.
&nbsp;  
At the time that \(??\{\}\) is compiled \(during execution of the entire
regexp\) I don't think the information that the whole regexp was
compiled under modifiers is available so this is not really surprising\.
But the resulting behaviour will probably surprise people anyway\.
&nbsp;   
Do we need a \(??imsx\-imsx\{\}\) shorthand for
\(??\{"\(?imsx\-imsx&#8203;:" \. RE \. "\)"\}\) that we can convert the above nodes to
at regexp compilation?
&nbsp;    
Or do we just hackishly convert them to \(??\{"\(?imsx\-imsx&#8203;:" \. RE \. "\)"\}\)?
&nbsp;     
Or do we document that modifiers aren't visible inside \(??\{\}\)?

\-\- 
Rick Delaney
rick@&#8203;consumercontact\.com


---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl v5\.6\.0&#8203;:

Configured by rick at Wed Apr 26 16&#8203;:08&#8203;:10 EDT 2000\.

Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.2\.10\, archname=i586\-linux
&nbsp;    uname='linux pgress\.consumercontact\.com 2\.2\.10 \#1 smp tue aug 10 19&#8203;:01&#8203;:45 mdt 1999 i586 unknown '
&nbsp;    config\_args='\-Doptimize=\-g'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define 
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-g'\, gccversion=egcs\-2\.91\.66 19990314 \(egcs\-1\.1\.2 release\)
&nbsp;    cppflags='\-DDEBUGGING \-fno\-strict\-aliasing'
&nbsp;    ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt
&nbsp;    libc=/lib/libc\-2\.1\.1\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.6\.0&#8203;:
&nbsp;    /usr/local/lib/perl5/5\.6\.0/i586\-linux
&nbsp;    /usr/local/lib/perl5/5\.6\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.6\.0/i586\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl v5\.6\.0&#8203;:
&nbsp;    HOME=/home1/rick
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/bin&#8203;:/usr/bin&#8203;:/usr/local/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From domo@ieee.org

> \[rick@&#8203;consumercontact\.com \- Mon Mar 12 00&#8203;:45&#8203;:07 2001\]&#8203;:

> $ perl \-le 'print "not ok" if "a\\n"\!~/\.\(??\{"\."\}\)/s'
> not ok
> $ perl \-le 'print "not ok" if "aA"\!~/a\(??\{"a"\}\)/i'
> not ok
>  
> etc\.

Still the case with 5\.8\.8 and bleadperl@&#8203;27631
>   
> At the time that \(??\{\}\) is compiled \(during execution of the entire
> regexp\) I don't think the information that the whole regexp was
> compiled under modifiers is available so this is not really surprising\.
> But the resulting behaviour will probably surprise people anyway\.
>    
> Do we need a \(??imsx\-imsx\{\}\) shorthand for
> \(??\{"\(?imsx\-imsx&#8203;:" \. RE \. "\)"\}\) that we can convert the above nodes to
> at regexp compilation?
>     
> Or do we just hackishly convert them to \(??\{"\(?imsx\-imsx&#8203;:" \. RE \. "\)"\}\)?
>      
> Or do we document that modifiers aren't visible inside \(??\{\}\)?
> 
Well\, that last appeals most to my laziness\, so I attach a documentation patch\, which cribs 
from the bug report\.

BTW\, \(??\{ and various other constructs have been marked as "highly experimental" since 5\.6\.1 
or earlier\. May they now be considered to be frozen and safe?

**Comment by p5pRT** [comment]
### From domo@ieee.org

<details><summary>perlre_postponed_modifiers.patch</summary>

```diff
--- perl-current-reference/pod/perlre.pod	2006-02-01 20:31:03.000000000 +0100
+++ perl-current-working/pod/perlre.pod	2006-03-30 15:01:40.000000000 +0200
@@ -724,6 +724,21 @@
 	     \)
 	  }x;
 
+Note that, by the time C<code> is evaluated, information about the modifiers
+(C<imsx>) that were in force when surrounding regular expression was evaluated
+is no longer available. Consequently, if a postponed regular expression
+requires modifiers, it must explicitly state them.
+
+The following regular expression does not match C<AA> because the C</i>
+modifier is not visible when the postponed part is evaluated:
+
+  /a(??{"a"})/i
+
+This issue may be worked around by including an explicit modifier in the
+postponed regular expression:
+
+  /a(??{"(?i:a)"})/
+
 =item C<< (?>pattern) >>
 X<backtrack> X<backtracking>
 

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

p5pers&#8203;:
this rather old ticket from Rick Delaney points out that in the following&#8203;:

&nbsp;    $some\_pattern\_string = 'A';
&nbsp;    "aa" =~ /A\(??\{ $some\_pattern\_string \}\)/i

the returned pattern string is compiled \*without\* /i \(or other such
flags\)\, and thus fails to match\.

The question is\, whether this behaviour is right\,and the docs need
clarifying\, or whether it should be "fixed"\.

I lean towards not\-a\-bug and fix the docs\, by analogy with

&nbsp;    my $some\_pattern = qr/A/; \# this doesn't use /i
&nbsp;    "aa" =~ /A\(??\{ $some\_pattern \}\)/i

which \(correctly IMHO\) fails to match\.

\-\- 
"Strange women lying in ponds distributing swords is no basis for a system
of government\. Supreme executive power derives from a mandate from the
masses\, not from some farcical aquatic ceremony\."
&nbsp;    \-\- Dennis\, "Monty Python and the Holy Grail"

**Comment by p5pRT** [comment]
### From @Leont

On Wed\, Jun 6\, 2012 at 12&#8203;:20 AM\, Leon Timmermans \<fawaka@&#8203;gmail\.com> wrote&#8203;:
> On Wed\, Jun 6\, 2012 at 12&#8203;:07 AM\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
>> p5pers&#8203;:
>> this rather old ticket from Rick Delaney points out that in the following&#8203;:
>>
>>    $some\_pattern\_string = 'A';
>>    "aa" =~ /A\(??\{ $some\_pattern\_string \}\)/i
>>
>> the returned pattern string is compiled \*without\* /i \(or other such
>> flags\)\, and thus fails to match\.
>>
>> The question is\, whether this behaviour is right\,and the docs need
>> clarifying\, or whether it should be "fixed"\.
>>
>> I lean towards not\-a\-bug and fix the docs\, by analogy with
>>
>>    my $some\_pattern = qr/A/; \# this doesn't use /i
>>    "aa" =~ /A\(??\{ $some\_pattern \}\)/i
>>
>> which \(correctly IMHO\) fails to match\.
>
> I think /\(??\{ $some\_pattern\_string \}\)/ should do the same as
> /$some\_pattern\_string/\, so I would see this as a bug\.
>
> Leon

And now to the correct address…

**Comment by p5pRT** [comment]
### From @rjbs

\* Dave Mitchell \<davem@&#8203;iabyn\.com> \[2012\-06\-05T18&#8203;:07&#8203;:37\]
> I lean towards not\-a\-bug and fix the docs\, by analogy with
> 
>     my $some\_pattern = qr/A/; \# this doesn't use /i
>     "aa" =~ /A\(??\{ $some\_pattern \}\)/i
> 
> which \(correctly IMHO\) fails to match\.

\.\.\.but the qr object in $some\_pattern implicitly carries its flags with it\, and
when stringified\, they're maintained\.  That is\, $some\_pattern isn't just the
string 'A' but '\(?^u&#8203;:A\)' \(or something very like that\)\.

So\, I don't think the analogy holds\.

Now\, that said\, I think it is still plausible to declare that the current
behavior is the preferred behavior\.  The kind of isolation provided by
\(??\{\.\.\.\}\) from other matching state makes me think I wouldn't be too confused
by finding that the flags on the matching do not affect it\, nor do I think it
would be too great a barrier to teaching the feature\.

I'm interested in what other people think\, though\!

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @rjbs

\* Leon Timmermans \<fawaka@&#8203;gmail\.com> \[2012\-06\-05T18&#8203;:24&#8203;:20\]
> On Wed\, Jun 6\, 2012 at 12&#8203;:20 AM\, Leon Timmermans \<fawaka@&#8203;gmail\.com> wrote&#8203;:
> >
> > I think /\(??\{ $some\_pattern\_string \}\)/ should do the same as
> > /$some\_pattern\_string/\, so I would see this as a bug\.

Do you think this is also a bug?

&nbsp;  use 5\.16\.0;
&nbsp;  my $str = 'foo';

&nbsp;  my $sub\_re = '\(f\)';

&nbsp;  $str =~ /\(??\{ $sub\_re \}\)/;
&nbsp;  my $lazy\_1 = $1;

&nbsp;  $str =~ /$sub\_re/;
&nbsp;  my $eager\_1 = $1;

&nbsp;  say "LAZY1&#8203;: $lazy\_1";
&nbsp;  say "EAGER&#8203;: $eager\_1";

Produces&#8203;:

&nbsp;  LAZY1&#8203;: 
&nbsp;  EAGER&#8203;: f

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Jun 05 15&#8203;:25&#8203;:11 2012\, LeonT wrote&#8203;:
> On Wed\, Jun 6\, 2012 at 12&#8203;:20 AM\, Leon Timmermans \<fawaka@&#8203;gmail\.com>
> wrote&#8203;:
> > On Wed\, Jun 6\, 2012 at 12&#8203;:07 AM\, Dave Mitchell \<davem@&#8203;iabyn\.com>
> wrote&#8203;:
> >> p5pers&#8203;:
> >> this rather old ticket from Rick Delaney points out that in the
> following&#8203;:
> >>
> >>    $some\_pattern\_string = 'A';
> >>    "aa" =~ /A\(??\{ $some\_pattern\_string \}\)/i
> >>
> >> the returned pattern string is compiled \*without\* /i \(or other such
> >> flags\)\, and thus fails to match\.
> >>
> >> The question is\, whether this behaviour is right\,and the docs need
> >> clarifying\, or whether it should be "fixed"\.
> >>
> >> I lean towards not\-a\-bug and fix the docs\, by analogy with
> >>
> >>    my $some\_pattern = qr/A/; \# this doesn't use /i
> >>    "aa" =~ /A\(??\{ $some\_pattern \}\)/i
> >>
> >> which \(correctly IMHO\) fails to match\.
> >
> > I think /\(??\{ $some\_pattern\_string \}\)/ should do the same as
> > /$some\_pattern\_string/\, so I would see this as a bug\.
> >
> > Leon

Perhaps surprisingly\, I have no opinion as to whether $some\_pattern
should inherit /i from the outer pattern\.

However\, if it doesn’t\, then ‘use re "/flags"’ should definitely apply\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Jun 05 15&#8203;:25&#8203;:11 2012\, LeonT wrote&#8203;:
> On Wed\, Jun 6\, 2012 at 12&#8203;:20 AM\, Leon Timmermans \<fawaka@&#8203;gmail\.com>
> wrote&#8203;:
> > On Wed\, Jun 6\, 2012 at 12&#8203;:07 AM\, Dave Mitchell \<davem@&#8203;iabyn\.com>
> wrote&#8203;:
> >> p5pers&#8203;:
> >> this rather old ticket from Rick Delaney points out that in the
> following&#8203;:
> >>
> >>    $some\_pattern\_string = 'A';
> >>    "aa" =~ /A\(??\{ $some\_pattern\_string \}\)/i
> >>
> >> the returned pattern string is compiled \*without\* /i \(or other such
> >> flags\)\, and thus fails to match\.
> >>
> >> The question is\, whether this behaviour is right\,and the docs need
> >> clarifying\, or whether it should be "fixed"\.
> >>
> >> I lean towards not\-a\-bug and fix the docs\, by analogy with
> >>
> >>    my $some\_pattern = qr/A/; \# this doesn't use /i
> >>    "aa" =~ /A\(??\{ $some\_pattern \}\)/i
> >>
> >> which \(correctly IMHO\) fails to match\.
> >
> > I think /\(??\{ $some\_pattern\_string \}\)/ should do the same as
> > /$some\_pattern\_string/\, so I would see this as a bug\.
> >
> > Leon

Perhaps surprisingly\, I have no opinion as to whether
$some\_pattern\_string should inherit /i from the outer pattern\.

However\, if it doesn‘t\, then ‘use re "/flags"’ should definitely apply\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @demerphq

On 6 June 2012 00&#8203;:07\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> p5pers&#8203;:
> this rather old ticket from Rick Delaney points out that in the following&#8203;:
>
>    $some\_pattern\_string = 'A';
>    "aa" =~ /A\(??\{ $some\_pattern\_string \}\)/i
>
> the returned pattern string is compiled \*without\* /i \(or other such
> flags\)\, and thus fails to match\.
>
> The question is\, whether this behaviour is right\,and the docs need
> clarifying\, or whether it should be "fixed"\.
>
> I lean towards not\-a\-bug and fix the docs\, by analogy with
>
>    my $some\_pattern = qr/A/; \# this doesn't use /i
>    "aa" =~ /A\(??\{ $some\_pattern \}\)/i
>
> which \(correctly IMHO\) fails to match\.

I think the documentation of \(??\{ \.\.\. \}\) says this should match\.

"The result of evaluation is considered as a regular expression and
matched as if it were inserted instead of this construct\."

Which says to me the above should match as /AA/i

Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 6 June 2012 00&#8203;:33\, Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> wrote&#8203;:
> \* Leon Timmermans \<fawaka@&#8203;gmail\.com> \[2012\-06\-05T18&#8203;:24&#8203;:20\]
>> On Wed\, Jun 6\, 2012 at 12&#8203;:20 AM\, Leon Timmermans \<fawaka@&#8203;gmail\.com> wrote&#8203;:
>> >
>> > I think /\(??\{ $some\_pattern\_string \}\)/ should do the same as
>> > /$some\_pattern\_string/\, so I would see this as a bug\.
>
> Do you think this is also a bug?
>
\[snip example of capture buffers not working\]

Note the docs say "and \*matched\* as if it were inserted instead of
this construct\." \(emphasis added\)\, it does not say "behaves exactly
the same as"\, or "captures the same thing as"\. And indeed it is
documented as such since at least 6bda09f9\.

Yves




\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @ap

\* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-06\-06 00&#8203;:30\]&#8203;:
> I'm interested in what other people think\, though\!

I think if a string is returned it should be compiled with the outer
pattern’s flags\, simply because of the relative difficulties of asking
for the opposite behaviour\.

Saying “disregard the outer pattern’s flags” explicitly is already very
easy to do by returning a qr// object\.

Now how would you go about saying “inherit the outer pattern’s flags”
explicitly? I \*think\* I can think of a way by arranging to compile
the pattern in stages via an outer qr// object and referring to its
stringification within the \(??\{\}\) construct and then pasting together
a new qr// object with that\, but gosh is that ever complicated – and
I don’t know if it’ll even work\.

Whereas if we say that strings inherit the outer pattern’s flags and
qr// objects don’t\, then it it’s easy for the user to specify which
behaviour is their preference\.

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
### From @rjbs

\* Aristotle Pagaltzis \<pagaltzis@&#8203;gmx\.de> \[2012\-06\-06T14&#8203;:25&#8203;:18\]
> \* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-06\-06 00&#8203;:30\]&#8203;:
> > I'm interested in what other people think\, though\!
> 
> I think if a string is returned it should be compiled with the outer
> pattern’s flags\, simply because of the relative difficulties of asking
> for the opposite behaviour\.

I am convinced\!

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jun 06\, 2012 at 08&#8203;:25&#8203;:18PM \+0200\, Aristotle Pagaltzis wrote&#8203;:
> \* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-06\-06 00&#8203;:30\]&#8203;:
> > I'm interested in what other people think\, though\!
> 
> I think if a string is returned it should be compiled with the outer
> pattern’s flags\, simply because of the relative difficulties of asking
> for the opposite behaviour\.

Ok I'm sold\. I've fixed this in my re\_eval branch\.

\-\- 
"But Sidley Park is already a picture\, and a most amiable picture too\.
The slopes are green and gentle\. The trees are companionably grouped at
intervals that show them to advantage\. The rill is a serpentine ribbon
unwound from the lake peaceably contained by meadows on which the right
amount of sheep are tastefully arranged\." \-\- Lady Croom\, "Arcadia"

**Comment by p5pRT** [comment]
### From @demerphq

On 6 June 2012 23&#8203;:58\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> On Wed\, Jun 06\, 2012 at 08&#8203;:25&#8203;:18PM \+0200\, Aristotle Pagaltzis wrote&#8203;:
>> \* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-06\-06 00&#8203;:30\]&#8203;:
>> > I'm interested in what other people think\, though\!
>>
>> I think if a string is returned it should be compiled with the outer
>> pattern’s flags\, simply because of the relative difficulties of asking
>> for the opposite behaviour\.
>
> Ok I'm sold\. I've fixed this in my re\_eval branch\.

Unrelated to this thread\, but related to your work\, is that I have in
smoke\-me a branch/patch to optimise NOTHING regops properly\. I mention
this because I noticed a lot of tests in pat\_re\_eval\.t start with

/\(?&#8203;:|\)\.\.\.\./

which my patch will optimize away entirely \(as if they weren't in the
pattern at all in this case\)\. Do you know what purpose these serve in
the tests?

Yves
\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 7 June 2012 08&#8203;:15\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> On 6 June 2012 23&#8203;:58\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
>> On Wed\, Jun 06\, 2012 at 08&#8203;:25&#8203;:18PM \+0200\, Aristotle Pagaltzis wrote&#8203;:
>>> \* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-06\-06 00&#8203;:30\]&#8203;:
>>> > I'm interested in what other people think\, though\!
>>>
>>> I think if a string is returned it should be compiled with the outer
>>> pattern’s flags\, simply because of the relative difficulties of asking
>>> for the opposite behaviour\.
>>
>> Ok I'm sold\. I've fixed this in my re\_eval branch\.
>
> Unrelated to this thread\, but related to your work\, is that I have in
> smoke\-me a branch/patch to optimise NOTHING regops properly\. I mention
> this because I noticed a lot of tests in pat\_re\_eval\.t start with
>
> /\(?&#8203;:|\)\.\.\.\./
>
> which my patch will optimize away entirely \(as if they weren't in the
> pattern at all in this case\)\. Do you know what purpose these serve in
> the tests?

Just pushed as&#8203;:

commit 3b6759a6b10290b6fea26fb98b50fbf5ce4466b5
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Wed Jun 6 08&#8203;:53&#8203;:05 2012 \+0200

&nbsp;    optimise \(?&#8203;:|\) and related NOTHING like constructs out of the
compiled optree

&nbsp;    A pattern like \(?&#8203;:|\) causes the regex engine to do extra work even though
&nbsp;    it is equivelent to matching nothing\. This optimises such sequences into
&nbsp;    more efficient opcodes that do less work\, and in some cases optimises
&nbsp;    them away entirely\.

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Jun 07\, 2012 at 08&#8203;:15&#8203;:49AM \+0200\, demerphq wrote&#8203;:
> On 6 June 2012 23&#8203;:58\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> > On Wed\, Jun 06\, 2012 at 08&#8203;:25&#8203;:18PM \+0200\, Aristotle Pagaltzis wrote&#8203;:
> >> \* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-06\-06 00&#8203;:30\]&#8203;:
> >> > I'm interested in what other people think\, though\!
> >>
> >> I think if a string is returned it should be compiled with the outer
> >> pattern’s flags\, simply because of the relative difficulties of asking
> >> for the opposite behaviour\.
> >
> > Ok I'm sold\. I've fixed this in my re\_eval branch\.
> 
> Unrelated to this thread\, but related to your work\, is that I have in
> smoke\-me a branch/patch to optimise NOTHING regops properly\. I mention
> this because I noticed a lot of tests in pat\_re\_eval\.t start with
> 
> /\(?&#8203;:|\)\.\.\.\./
> 
> which my patch will optimize away entirely \(as if they weren't in the
> pattern at all in this case\)\. Do you know what purpose these serve in
> the tests?

No real idea\. They were added by Bram to test the behaviour of $^N and $\+ in
\(?\{\}\)\. So I guess he was trying to test all the permutations of
backtracking and paren index restoring\.

\-\- 
"But Sidley Park is already a picture\, and a most amiable picture too\.
The slopes are green and gentle\. The trees are companionably grouped at
intervals that show them to advantage\. The rill is a serpentine ribbon
unwound from the lake peaceably contained by meadows on which the right
amount of sheep are tastefully arranged\." \-\- Lady Croom\, "Arcadia"

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jun 06\, 2012 at 10&#8203;:58&#8203;:37PM \+0100\, Dave Mitchell wrote&#8203;:
> On Wed\, Jun 06\, 2012 at 08&#8203;:25&#8203;:18PM \+0200\, Aristotle Pagaltzis wrote&#8203;:
> > \* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-06\-06 00&#8203;:30\]&#8203;:
> > > I'm interested in what other people think\, though\!
> > 
> > I think if a string is returned it should be compiled with the outer
> > pattern’s flags\, simply because of the relative difficulties of asking
> > for the opposite behaviour\.
> 
> Ok I'm sold\. I've fixed this in my re\_eval branch\.

\(but seeing tests fail in regexp\_qr\_embed\.t\)\, a further question&#8203;:

should \(?flags&#8203;:\.\.\.\) also propagate into \(??\{\}\)\, e\.g\. should this match&#8203;:

&nbsp;    print "match\\n" if "A" =~ /^\(?i&#8203;:\(??\{'a'\}\)\)/;

I noted this because regexp\_qr\_embed\.t does

&nbsp;    my $RE = qr/\(??\{'a'\}\)/i;
&nbsp;    "A" =~ /\(&#8203;:\)$RE\(?&#8203;:\)/

which in my code is failing\, since I'm propagating the /i flag but not
the \(?i&#8203;:\) flag\.


\-\- 
"Do not dabble in paradox\, Edward\, it puts you in danger of fortuitous wit\."
&nbsp;    \-\- Lady Croom\, "Arcadia"

**Comment by p5pRT** [comment]
### From @demerphq

On 7 June 2012 10&#8203;:39\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> On Wed\, Jun 06\, 2012 at 10&#8203;:58&#8203;:37PM \+0100\, Dave Mitchell wrote&#8203;:
>> On Wed\, Jun 06\, 2012 at 08&#8203;:25&#8203;:18PM \+0200\, Aristotle Pagaltzis wrote&#8203;:
>> > \* Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> \[2012\-06\-06 00&#8203;:30\]&#8203;:
>> > > I'm interested in what other people think\, though\!
>> >
>> > I think if a string is returned it should be compiled with the outer
>> > pattern’s flags\, simply because of the relative difficulties of asking
>> > for the opposite behaviour\.
>>
>> Ok I'm sold\. I've fixed this in my re\_eval branch\.
>
> \(but seeing tests fail in regexp\_qr\_embed\.t\)\, a further question&#8203;:
>
> should \(?flags&#8203;:\.\.\.\) also propagate into \(??\{\}\)\, e\.g\. should this match&#8203;:
>
>    print "match\\n" if "A" =~ /^\(?i&#8203;:\(??\{'a'\}\)\)/;
>
> I noted this because regexp\_qr\_embed\.t does
>
>    my $RE = qr/\(??\{'a'\}\)/i;
>    "A" =~ /\(&#8203;:\)$RE\(?&#8203;:\)/
>
> which in my code is failing\, since I'm propagating the /i flag but not
> the \(?i&#8203;:\) flag\.

imo \(assume /x\)&#8203;:

\(?&#8203;: \(?i\) foo \)

and

\(?i&#8203;: foo \)

should produce the same results\, so

\(?&#8203;: \(?i\) \(??\{ "foo" \}\) \)

shuld be the same as&#8203;:

\(?i&#8203;: \(??\{"foo"\}\) \)


cheers\,
Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by ec841a2713d\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
